### PR TITLE
DNM: fix: Clients now correctly quote/unquote ref parts

### PR DESCRIPTION
**Pull request by tssweeney**
**Summary:** This PR hopefully 🤞 solves >90% of our remaining Ref/Naming issues - meaning users should be able to freely create entities/projects (as long as Gorilla accepts them) and name their objects/ops anything they want (baring a `%` symbol). Under the hood, I close a few remaining constraints / parsing issues on the server side and mostly modify the UI and Python clients to correctly read/write quoted refs.

---

High Level Improvements:
1. Python Client will always produce valid refs (ie. it correctly quotes when needed)
   * Benefit to user: no more crashed python scripts when bad chars are in the name
   * Note: the one case of `%` in the name is still illegal. We sanitize on the python client for this specific case.
2. Python Client is able to consume quoted refs (ie. it correctly unquotes when needed)
   * Benefit to user: able to correctly walk ref data edges with bad chars in the name
3. UI Client will always produce valid refs (ie. it correctly encodes when needed)
   * Benefit to user: able to link safely when dealing with bad characters
4. UI Client is able to consume encoded refs (ie. it correctly decodes when needed)
   * Benefit to user: able to correctly display links with bad characters

Smaller Details:
* Adds correct quoting of URLs to printing URLs
* Adds server-side entity/project de-quoting

Companion PR: https://github.com/wandb/core/pull/23333
* Correctly routes when parts are encoded
* Correctly de-quoting encoded refs on project lookup

---

This PR finishes up the Ref work to ensure we support any ref the user throws at us. 

**Context:**
In a recent PR: https://github.com/wandb/weave/pull/2077, I implemented a more robust parser and checker to ensure Refs and Names conform to expectations. This stopped the "bleeding" where we were storing invalid Refs.

However, our two clients Python and UI had not been updated to reflect the new rules. In fact, the Python client was still creating bad refs, and the UI was overly restrictive with what it thought a ref could be. So double edge sword. As a result, users still hit crashes (either python write time or ui read time).

Moreover, our Ref parsing logic is a mess (especially in the UI) as it is spread across many locations & riddled with legacy junk.

*A little terminology note*: in PY, URL escaping is called `quoting` and in TS it is `encoding`. When used in this PR, they follow the language-specific convention, but logically mean the same concept.

**Understanding Refs:**
See https://www.notion.so/wandbai/WIP-Robust-Refs-550bf06f4a244a1ab4fdd26133952e2c for more details.

The important part to understand is the ref format:

```
Calls:
    weave:///<entity>/<project>/call/<callid>[extra]
Ops:
    weave:///<entity>/<project>/op/<name>:<digest>[extra]
Objects:
    weave:///<entity>/<project>/object/<name>:<digest>[extra]
Tables:
    weave:///<entity>/<project>/table/<digest>[extra]
Where:
[extra] is an optional set of slash-seperated tuples of edges. for example:
    /key/key_name_1/index/0
They always follow the pattern of [edgeName/edgeValue], where 
`edgeName` is a shortlist of 4 specific edges names.
```

Furthermore, `entity`, `project`, `name` and the `edgeValue`s are user-controlled.

The main constraints we need to parse these refs are:
1. `entity` cannot contain slashes `/`
2. `project` cannot contain slashes `/`
3. `name` cannot contain slashes `/` or colons `:`
4. `edgeValue` cannot contain slashes `/`

Therefore, we need to basically URL encode/decode each of these parts in all the places that they are used

* This PR basically updates the parsing and construction logic to properly encode/decode `entity`, `project`, name` and `edgeValue`.
* This adds server-side decoding of entity/project per this more rigorous spec

---

**Deployment**
The ideal order would be:
1. The companion PR (land and deploy): https://github.com/wandb/core/pull/23333
2. Secondly, split out the trace server changes from this PR (land and deploy)
3. Then we can land & deploy/release UI and python changes in any order
4. Make sure prod is good for few days
5. Rollout to dedicated

### CODE REVIEW

File: weave/trace_server/refs_internal.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class InvalidInternalRef(ValueError):
     pass
 
 
-def quote_select(s: str, quote_chars: tuple[str, ...] = ("/", ":", "%")) -> str:
+def quote_select(s: str, quote_chars: tuple[str, ...] = ("%", "/", ":")) -> str:
```
Comment (tssweeney) [inline weave/trace_server/refs_internal.py:32]: This is a legit bug fix!

File: weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export const flattenObjectPreservingWeaveTypes = (obj: {
   [key: string]: any;
 }) => {
   return flattenObject(obj, '', {}, (key, value) => {
-    return typeof value !== 'object' || value == null || value._type == null;
+    return typeof value !== 'object' || value == null || value._type !== 'CustomWeaveType';
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts:10]: This is an unrelated fix I stumbled upon while testing

File: weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts [hunk 3]
```diff
@@ -7,7 +7,11 @@ export const flattenObjectPreservingWeaveTypes = (obj: {
   [key: string]: any;
 }) => {
   return flattenObject(obj, '', {}, (key, value) => {
-    return typeof value !== 'object' || value == null || value._type == null;
+    return (
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts:10]: This is a little followup fix from the Image PR i just put in - it is unrelated to refs

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/utilities.ts [hunk 4]
```diff
@@ -48,8 +49,13 @@ export const refUriToOpVersionKey = (refUri: RefUri): OpVersionKey => {
 };
 
 export const opVersionKeyToRefUri = (key: OpVersionKey): RefUri => {
-  return `${WEAVE_REF_PREFIX}${key.entity}/${key.project}/op/${key.opId}:${key.versionHash}`;
-  // return `${WANDB_ARTIFACT_REF_PREFIX}${key.entity}/${key.project}/${key.opId}:${key.versionHash}/obj`;
+  return makeRefObject(
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/utilities.ts:52]: We have too much ref parsing logic floating around, centralizing to one location.

File: weave-js/src/components/PagePanelComponents/Home/Browse3/refs.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/refs.ts:1]: much of this file already existed - moved from `weave-js/util/refs.ts` to here as this is scoped to "New Weave" (aka Browse3)

File: weave-js/src/util/refs.ts
Comment (tssweeney) [inline weave-js/src/util/refs.ts:1]: Moved to `weave-js/src/components/PagePanelComponents/Home/Browse3/refs.ts`

File: weave/init_message.py [hunk 6]
```diff
@@ -44,10 +44,17 @@ def _print_version_check() -> None:
             if use_message:
                 print(use_message)
 
-    orig_module = wandb._wandb_module
-    wandb._wandb_module = "weave"
-    weave_messages = wandb.sdk.internal.update.check_available(weave.__version__)
-    wandb._wandb_module = orig_module
+    weave_messages = None
```
Comment (tssweeney) [inline weave/init_message.py:47]: This is unrelated to Refs... for some reason mypy was complaining about this.

File: weave/trace/refs.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import dataclasses
```
Comment (tssweeney) [inline weave/trace/refs.py:1]: Major changes in this file are to ensure that we quote the user-controlled fields, similar to refs_internal

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 8]
```diff
@@ -1404,15 +1405,6 @@ const isValidTraceCall = (callRes: traceServerTypes.TraceCallSchema) => {
...
 
-export const privateRefToSimpleName = (ref: string) => {
-  const trimmed = ref.replace(`${WEAVE_PRIVATE_PREFIX}//`, '');
-  try {
-    return trimmed.split('/')[1].split(':')[0];
-  } catch (e) {
-    return trimmed;
-  }
-};
-
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:1415]: nice

File: weave-js/src/react.tsx [hunk 9]
```diff
@@ -501,10 +502,29 @@ export const isWeaveObjectRef = (ref: ObjectRef): ref is WeaveObjectRef => {
...
 };
 
+const trimExtra = (extra: string | undefined) => {
+  if (extra?.startsWith('/')) {
+    extra = extra.slice(1);
+  }
+  if (extra?.endsWith('/')) {
+    extra = extra.slice(0, -1);
+  }
+  if (extra === '') {
```
Comment (gtarpenning) [inline weave-js/src/react.tsx:512]: could also do `extra.length === 0`, its faster 🤷🏻

File: weave-js/src/react.tsx [hunk 10]
```diff
@@ -614,7 +632,7 @@ export const parseRef = (ref: string): ObjectRef => {
         artifactRefExtra: '',
```
Comment (gtarpenning) [inline weave-js/src/react.tsx:632]:
```suggestion
        artifactRefExtra: undefined,
```

Comment (gtarpenning) [reply weave-js/src/react.tsx:632]: if we want to be consistent

File: weave-js/src/react.tsx [hunk 11]
```diff
@@ -627,7 +645,7 @@ export const parseRef = (ref: string): ObjectRef => {
...
```
Comment (gtarpenning) [inline weave-js/src/react.tsx:645]: +1

**Review by gtarpenning — APPROVED** [review]
lgtm, but i know @jamie-rasmussen had good thoughts on encoding in the frontend, the last time i tried to meddle with it

File: weave-js/src/components/PagePanelComponents/Home/Browse3/NotFoundPanel.tsx [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import {ErrorPanel} from '@wandb/weave/components/ErrorPanel';
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/NotFoundPanel.tsx:1]: I noticed while testing that it is pretty lame when a not found error crashes the whole page. This is a common component that can be used to render not found data

File: weave-js/src/components/PagePanelComponents/Home/Browse3/context.tsx [hunk 13]
```diff
@@ -92,8 +92,13 @@ export const browse2Context = {
   entityUrl: (entityName: string) => {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/context.tsx:92]: Pretty much this whole file is all about encoding the url parts

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/CallPage.tsx [hunk 14]
```diff
@@ -45,16 +40,7 @@ export const CallPage: FC<{
...
-      <div style={{display: 'flex', flexDirection: 'column', height: '100%'}}>
-        <div style={{alignSelf: 'flex-end', margin: 10}}>
-          <Button icon="close" variant="ghost" onClick={close} />
-        </div>
-        <div style={{flex: 1}}>
-          <ErrorPanel title="Call not found" subtitle="" subtitle2="" />
-        </div>
-      </div>
-    );
+    return <NotFoundPanel title="Call not found" />;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/CallPage.tsx:43]: Moved to a common component

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 15]
```diff
@@ -669,6 +670,13 @@ const useOpVersion = (
       };
     }
 
+    if (opVersionRes.obj == null) {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:673]: This block (and the one below that is similar basically handles the case when the target data is not found. before, this would create a JS error due to incorrect nulls

File: weave-js/src/react.tsx [hunk 16]
```diff
@@ -599,71 +620,74 @@ export const parseRef = (ref: string): ObjectRef => {
...
-    ] = match.slice(1);
+  throw new Error(`Unknown protocol: ${url.protocol}`);
+};
+
+/**
+ * This is the "New Weave" ref parser that doesn't need to worry about W&B
+ * artifacts or local artifacts. It is not exported currently, but is used for
+ * the new Weave branch in parseRef
+ */
+const parseWeaveTraceRef = (ref: string): WeaveObjectRef => {
```
Comment (tssweeney) [inline weave-js/src/react.tsx:631]: This was extracted from `parseRef` and is easier to review with whitespace hidden

File: weave-js/src/components/PagePanelComponents/Home/Browse3/context.tsx [hunk 17]
```diff
@@ -428,7 +475,11 @@ type RouteType = {
...
   entityUrl: (entityName: string) => string;
-  projectUrl: (entityName: string, projectName: string) => string;
+  projectUrl: (
+    entityName: string,
+    projectName: string,
+    noEncode?: boolean
+  ) => string;
   typeUIUrl: (
     entityName: string,
     projectName: string,
```
Comment (vanpelt) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/context.tsx:485]: Can't we assume entityName and projectName are [always safe](https://github.com/wandb/core/blob/master/services/gorilla/validate.go#L30)?

File: weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts [hunk 18]
```diff
@@ -7,7 +7,11 @@ export const flattenObjectPreservingWeaveTypes = (obj: {
   [key: string]: any;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse2/browse2Util.ts:7]: Moved to New PR

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/OpVersionPage.tsx [hunk 19]
```diff
@@ -35,7 +36,7 @@ export const OpVersionPage: React.FC<{
   if (opVersion.loading) {
     return <CenteredAnimatedLoader />;
   } else if (opVersion.result == null) {
-    return <div>Op version not found</div>;
+    return <NotFoundPanel title="Op not found" />;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/OpVersionPage.tsx:39]: moved to new PR

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/ObjectVersionPage.tsx [hunk 20]
```diff
@@ -59,7 +60,7 @@ export const ObjectVersionPage: React.FC<{
   if (objectVersion.loading) {
     return <CenteredAnimatedLoader />;
   } else if (objectVersion.result == null) {
-    return <div>Object not found</div>;
+    return <NotFoundPanel title="Object not found" />;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/ObjectVersionPage.tsx:63]: moved to new PR

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/CallPage.tsx (hunk 14 above)
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/CallPage.tsx:43]: moved to new PR

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 15 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:673]: moved to new PR
